hydro refactors

rgknox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description:

This PR contains a variety of changes and enhancements that try to make the plant hydraulics code more robust.  This should be synced with:   [ESCOMP/CTSM 929](https://github.com/ESCOMP/CTSM/pull/929)

---------------------------------
1) various state variables were pruned from memory on an audit of what is explicitly needed. 
2) The call sequence and how variables and information are transferred and stored (particularly conductances) was over-hauled (emphasis was placed on trying not to copy variables from one location to another, but rather calculate in-line with its usage, for clarity), 
3) the plant and soil water transfer functions, be it theta-psi, or fraction of total conductance, or their derivatives was converted to a modular environment, 
4) some modifications to diagnostics 
5) @pnlfang 's newton matrix solver was incorporated into this new framework (however, at the time of this writing, while it is fully written and has had limited testing, it is not yet operational), 
6) the partitioning of plant compartments was changed in a subtle way in that all compartments are now split in half at the node center, with an upstream and downstream side that has different material properties, 
7) a hard-coded option to ignore the top soil layer is now available (and on by default, ie it is ignored) 
8) all instances of undefined constants have been removed (ie no numbers in code that are not defined), 
9) the serial (original) solve was given an option to either parallelize in time or space, the default is now parallel in space, 
10) the units of the site-level rhizosphere shells are now m3/site (absolute units)  and 
11) lots of code cleaning
12) allowing the water retention functions to operate beyond saturation, thus enabling us to get away from forcing water states artificially below saturation

----------------------------

At present, this refactor generates reasonable and stable results at BCI and California CZ2. 

Remaining Tasks Before integration:

1) Code review:
2) Get tests passing in our standard test suite
3) More cleaning and document code
4) Review history variables (add/modify/change?)
5) Test grasses and shrubs
6) Deciduous plants have not been thoroughly tested (however the refactors did not target this behavior, so it may be fine)
7) Since the Newton Solve is optional, might push debugging that to another PR, depending on discussions with @pnlfang 

Remaining Long-Term Tasks:
Change state variable to compartment volume and water mass. This should simplify initialization and remove some variables.
Root layers spanning multiple soil layers.  Not top priority, and wouldn’t work until the newton solve is working.


<!--- Describe your changes in detail -->
<!--- please add issue number if one exists -->

### Collaborators:
@bchristo @xuchongang @pnlfang @ckoven @JunyanDing and others

<!--- List names of collaborators or people who have interacted -->
<!--- in bringing about this set of changes -->
<!--- consultation, discussions, etc. -->

### Expectation of Answer Changes:
<!--- Please describe under what conditions, if any, -->
<!--- the model is expected to generated different answers -->
<!--- from the master version of the code -->

yes, when hydraulics is on, this should change answers.  These changes are not the result of fundamental new hypotheses, but due to a litany of algorithmic changes

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [ ] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided


### Test Results:
<!--- Non-trivial changes require the PASS/FAIL regression tests. -->
<!--- If changes to code are NOT expected to change answers, tests must -->
<!--- be run against a baseline. -->

TBD

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

<!--- paste in test results here -->


<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

rgknox commented:
Here is some documentation. It is pretty rough right now, but might be helpful in the mean time if people want another description of how things work along with Brad's TFS paper:  https://drive.google.com/file/d/1GSl-TeT4n3j7CDNjNm2rRdGWg55KLWhQ/view?usp=sharing

rgknox commented:
Test Results at BCI:https://drive.google.com/file/d/1tuyBx5b0g_V2QAvgw6Ns6W4eqJNKSppJ/view?usp=sharing

Test Results at CZ2 (California): https://drive.google.com/file/d/17ZsPCoy9VHXn4n9OilXFbnon_8arUY63/view?usp=sharing

rgknox commented:
Tested gnu/intel.  All non-hydro tests match baseline.
Also, now hydro tests PASS:
/glade/scratch/rgknox/clmed-tests/fates-sci.1.34.0_api.9.0.0-clm5.0.30-C3fa9452e-F74aa9fcd.fates.cheyenne.intel

jenniferholm commented:
Ryan, do these hydro refactors work with ELM?

On Tue, Feb 18, 2020 at 9:48 PM Ryan Knox <notifications@github.com> wrote:

> Tested gnu/intel. All non-hydro tests match baseline.
> Also, now hydro tests PASS:
>
> /glade/scratch/rgknox/clmed-tests/fates-sci.1.34.0_api.9.0.0-clm5.0.30-C3fa9452e-F74aa9fcd.fates.cheyenne.intel
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NGEET/fates/pull/611?email_source=notifications&email_token=ACCVHGEKI4J4UV5HKVKTE6TRDTBZ3A5CNFSM4KXLFY6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMGOKVQ#issuecomment-588047702>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCVHGFDQZ6AHVANFJA7BBTRDTBZ3ANCNFSM4KXLFY6A>
> .
>


-- 
Jennifer Holm
Research Scientist
Climate and Ecosystems Sciences Division
Lawrence Berkeley National Laboratory
510-495-8083

rgknox commented:
Not yet, but the coding to make it happen should be straight forward.  I will create a compatible branch to use while we are reviewing this.

rgknox commented:
I'm refactoring the root uptake flux diagnostic.  Two questions, what are the preferable units (scaling) here and what dimensions are most useful?

As for units, we could have flux per individual [kg/indiv/s], or scale that flux by crown area? I suppose, we could scale it by crown area in the post-processing stage. I would assume flux per individual is only meaningful if the diagnostic is dimensioned by size....

As for dimensions, we could partition by plant size-class, pft, soil layer.  

One approach would be to generate a pft x size class diagnostic, but create a few history variables with different depth ranges like 0-10, 10-50, 50-100, 100+ cm depth?

@xuchongang @bchristo @pnlfang @tompowell9 @JunyanDing @pollybuotte @jenniferholm

JunyanDing commented:
Yes, I think set the unit by individual makes more sense than by crown area.

As for dimension, can we have pft x size, then use the same soil layers as defined in the surface data?

ckoven commented:
I'd argue to just sum it over cohorts to get as a straight mass flux in kg/s/ha. If you want to put in units of per individual, per unit crown area, etc, you can always divide it by either of those quantities in postprocessing.

not sure for dimensions...

pollybuotte commented:
I agree with summing over cohorts, and dimensions of PFT x size and soil layer.

rgknox commented:
Here is the ELM branch I modified for compatibility with the hydro-refactor:  https://github.com/rgknox/E3SM/tree/rgknox/lnd/fates-hydro-refactor
(this branch is untested at the time of this post)

rgknox commented:
The testing on this branch has progressed to the point where I invite others to test this out.
Make sure to use the branches (E3SM/CTSM) I indicate in the discussion to be compatible.
I don't plan anymore developments on this, and will leave as is, with the exception of addressing reviewers and community feedback.  The following have expressed interest in potentially testing and using this branch: @xuchongang , @bchristo, @rosiealice , @jenniferholm @JunyanDing.  All feedback, either by formal review or otherwise is appreciated.

I'm also planning on providing a walk-through of the code changes to potential code reviewers, please let me know if you are interested in participating.

JunyanDing commented:
Thanks Ryan!
I will start trying this new branch at CZ2 aftet next week's land model group meeting. Please keep posted on the code change walk-through. I definitely will benefit from this. I am really appreciated!

jenniferholm commented:
@rgknox - Since you are still working on the ELM interfacing, here is a small issue.
I've been doing testing on ELM with Dan's @dmricciuto  help and he found there are some Interface variables that were not updated or changed names but forgot to be included in clmfates_interfaceMod.F90, and as a result this causes the model to crash in a very similar to the transpiration and NaN issue we fixed awhile ago. 
https://github.com/NGEET/fates/pull/555

The last clmfates_interfaceMod.f90 was updated on Oct 25, 2019... and can be found here - https://github.com/E3SM-Project/E3SM/commit/464d1a60662fd20ca24ac7c7b622b1063d9c3e45#diff-26b8ffc0ec3b66f6b87a776e9ee64416

It looks like some water flux variables changed names (veg_wf%qflx_tran_veg(p)). Here are the changes that can be made, and afterwards our ELM runs were successful. 

--- a/components/clm/src/main/clmfates_interfaceMod.F90
+++ b/components/clm/src/main/clmfates_interfaceMod.F90

@@ -92,7 +92,7 @@ module CLMFatesInterfaceMod
    use TopounitDataType  , only : top_as
    use ColumnType        , only : col_pp
    use ColumnDataType    , only : col_es, col_ws, col_wf, col_cs, col_cf
-   use VegetationDataType, only : veg_es, veg_wf   
+   use VegetationDataType, only : veg_es, veg_wf, veg_ef  
    use LandunitType      , only : lun_pp

    use landunit_varcon   , only : istsoil

@@ -2340,7 +2340,8 @@ contains
       do ifp = 1, this%fates(nc)%sites(s)%youngest_patch%patchno 
          p = ifp+col_pp%pfti(c)
          this%fates(nc)%bc_in(s)%swrad_net_pa(ifp) = solarabs_inst%fsa_patch(p)
-         this%fates(nc)%bc_in(s)%lwrad_net_pa(ifp) = energyflux_inst%eflx_lwrad_net_patch(p)
+         this%fates(nc)%bc_in(s)%lwrad_net_pa(ifp) = veg_ef%eflx_lwrad_net(p)

       end do
    end do

@@ -2353,7 +2354,7 @@ contains
       c = veg_pp%column(p)
       s = this%f2hmap(nc)%hsites(c)
       ifp = p - col_pp%pfti(c)
-      this%fates(nc)%bc_in(s)%qflx_transp_pa(ifp) = waterflux_inst%qflx_tran_veg_patch(p)
+      this%fates(nc)%bc_in(s)%qflx_transp_pa(ifp) = veg_wf%qflx_tran_veg(p)

rgknox commented:
I'm testing some fixes that address issues brought up by @JunyanDing and @mariuslam

beta testers are welcome to try this branch:
https://github.com/rgknox/fates/tree/hydro-diagnostics-wmat-gfixes

If it looks promising, then I will merge in to this branch.  Here is the diff:

https://github.com/rgknox/fates/compare/hydro-diagnostics-wmat...rgknox:hydro-diagnostics-wmat-gfixes

rgknox commented:
@JunyanDing and I are both getting stability improvements with the new branch, will merge it in now

rgknox commented:
I will resolve conflicts on this, looks like they re-emerged with the size-age mortality stuff.

This branch has already benefited from beta testing.  The last phase is to give folks a last chance to do code review.

Calling all code reviewers!

rgknox commented:
Thanks @xuchongang ! I will take a look and address your comments

rgknox commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+    real(r8)      , intent(in)     :: pinot
+    real(r8)      , intent(out)    :: psi         ! water potential   [MPa]
+
+    psi = pinot*th_sat*(rwc_fd - th_res) / (th - th_sat*th_res)
```
@xuchongang @JunyanDing @bchristo :

I worked it through too, and I think I understand now.

I rearranged a little bit, it looks like all the terms in the numerator and denominator are meant to be in units that match saturation fraction.

So if units of resid matches saturation fraction units: 

y = pinot * (rwcft - resid) / ( x/thetas - resid) 


If resid is in water content units (absolute, or same as thetas and theta):

y = pinot * (rwcft - resid/thetas) / ( x/thetas - resid/thetas) 

   = pinot * (rwcft*thetas - resid) / (x-resid)


I'm fine with either unit, just let me know what you all prefer and I will work it out.  We should also update the parameter file, and @JunyanDing noted that the derivative needs fixing too I think.

rgknox replied:
I'm noting other places where we treat resid with units that are the same as water conent (not relativistic):

https://github.com/NGEET/fates/blob/c99c5bdf99cdb2ef6e8894b853b5b8ee9cb84627/biogeophys/FatesHydroWTFMod.F90#L823

bchristo replied:
I can confirm what Chonggang noted -- indeed, anywhere in the old code there was the variable `resid`, it was as a fraction of `thetas`

Poor nomenclature, I can see how this misinterpretation happened. But the implication here is that anywhere in the current refactor that `th_res` represents a replacement of `resid` then that should be fixed.

xuchongang replied:
@rgknox and @bchristo , I would prefer using th_res as the same unit of thetas and theta. We only need to update the equations of derivative and water content, where old resid used.

bchristo replied:
@rgknox and @xuchongang Here's my preference for how to deal with this:

- Change name of parameter in paramfile to rwc_r and note the relativistic RWC units there
- Calculate in `set_wrf_param_tfs` `th_res` as `thetas*rwc_r`
- Replace `th_res` in `solutepsi()`, `dpsolutepsi()`, `pressurepsi()`, and `dpressurepsi()` with `rwc_r`

rgknox replied:
We could change the name of the parameter in the file in a later pull request, just not on this PR, as it will create incompatabilities for other users.  I think we have a PR coming soon to update a group of parameters.  That being said, the current parameter name seems suitably ambiguous, at least for right now, and we could change the descriptive text (in this PR) to make it clear that is the fraction of th_sat.   I will go with whatever you and @xuchongang decide.

rgknox replied:
in an offline discussion, there was agreement that we will change the th_res parameter to have volumetric units, similar to th_sat

rgknox commented on biogeophys/FatesPlantHydraulicsMod.F90:
```diff
@@ -1014,246 +1037,178 @@ function constrain_water_contents(th_uncorr, delta, ft, k) result(th_corr)
-    ths     = EDPftvarcon_inst%hydr_thetas_node(ft,porous_media(k))
-    thr     = ths * EDPftvarcon_inst%hydr_resid_node(ft,porous_media(k))
+    ths     = EDPftvarcon_inst%hydr_thetas_node(ft,pm_type)
+    thr     = ths * EDPftvarcon_inst%hydr_resid_node(ft,pm_type)
```
I see, so lets make sure the meta-data makes it clear that hydr_resid_node is per m3 of maximum saturation, ie relative units, or relative to saturation.

xuchongang replied:
@rgknox , I think the current definition is in relative terms already. Thus it is OK as it is for me; however, it might be cleaner and not easier to be confused if we set hydr_resid_node the same unit as hydr_thetas_node.

rgknox replied:
I like those units more too, will change

JunyanDing commented:
Great! Didn't notice this one. I think these might be the causes of
unstable 1D solving.

Junyan


On Mon, Apr 6, 2020 at 3:23 PM Ryan Knox <notifications@github.com> wrote:

> *@rgknox* commented on this pull request.
> ------------------------------
>
> In biogeophys/FatesHydroWTFMod.F90
> <https://github.com/NGEET/fates/pull/611#discussion_r404424536>:
>
> > +    !
> +    ! !DESCRIPTION: computes solute water potential (negative) as a function of
> +    !  water content for the plant PV curve.
> +    !
> +    ! !USES:
> +    !
> +    ! !ARGUMENTS
> +
> +    real(r8)      , intent(in)     :: th          ! vol wc       [m3 m-3]
> +    real(r8)      , intent(in)     :: rwc_fd
> +    real(r8)      , intent(in)     :: th_sat
> +    real(r8)      , intent(in)     :: th_res
> +    real(r8)      , intent(in)     :: pinot
> +    real(r8)      , intent(out)    :: psi         ! water potential   [MPa]
> +
> +    psi = pinot*th_sat*(rwc_fd - th_res) / (th - th_sat*th_res)
>
> I'm noting other places where we treat resid with units that are the same
> as water conent (not relativistic):
>
>
> https://github.com/NGEET/fates/blob/c99c5bdf99cdb2ef6e8894b853b5b8ee9cb84627/biogeophys/FatesHydroWTFMod.F90#L823
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NGEET/fates/pull/611#discussion_r404424536>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKON6O7F7463TVEBXHXLZ3DRLJI6JANCNFSM4KXLFY6A>
> .
>

rgknox commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+  real(r8), parameter :: min_ftc = 0.0005_r8   ! Minimum allowed fraction of total conductance
+                                               
+
+  real(r8), parameter :: min_sf_interp = 0.02 ! Linear interpolation above this saturated frac
```
The intention here was to prevent flow conditions with exactly zero flow.  This number was intended to be suitably small to avoid the strangeness of zero's in math, but still be small enough to not really have an affect on results.  Note the units here are fraction of conductance, not saturation fraction. This also will not place a lower bound on saturated fraction. I suppose I could just remove this cap.  Thoughts @xuchongang ?

bchristo replied:
I had a similar comment. @rgknox , I think @xuchongang was referring to line 34, not line 31. He meant the comment should read "Linear interpolation below this saturated frac"

I like the min_ftc to avoid zero flow. I agree with your rationale.

rgknox commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+     real(r8) :: th_res   ! Residual volumentric water content          [m3/m3]
+     real(r8) :: pinot    ! osmotic potential at full turger            [MPa]
+     real(r8) :: epsil    ! bulk elastic modulus                        [MPa]
+     real(r8) :: rwc_fd   ! total RWC @ which elastic drainage begins     [-]
```
ahhh, I see, will switch name

bchristo replied:
Yes, agreed (rwc_ft) and 'ft' stands for 'full turgor'

rgknox commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+  function th_from_psi_base(this,psi) result(th)
+    class(wrf_type)     :: this
+    real(r8),intent(in) :: psi
+    real(r8)            :: th
```
It is implicitly intent out because it is defined in the result() operator.  I suppose we could label this as intent out to re-inforce the point. I will see if the compiler has issues with this.

rgknox replied:
I did check this and at least the gnu compilers don't allow us to specify the intent of the result, as the result of a function is implied to be out.

rgknox commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+
+  ! =====================================================================================
+
+  function th_from_psi_vg(this,psi) result(th)
```
I had originally intended to allow positive pressures in all the different functional hypotheses, not just the tfs style. If I can add the linear extrapolations to VG and CCH, will do, otherwise, I will add a failure clause.

bchristo replied:
@rgknox , Sounds good to me.

rgknox commented:
I think a useful diagnostic would be the mean water availability of the soil, weighted by root zone surface area for different layers.  

We could present this as either the volumetric water content, saturated fraction, or water potential.

It could be dimensioned by PFT and/or plant size.

Anyone like this idea, or have improvements?

bchristo commented on main/FatesInterfaceMod.F90:
```diff
@@ -574,7 +571,12 @@ module FatesInterfaceMod
+      real(r8),allocatable :: qflx_ro_sisl(:)          ! Water flux runoff generated by
+                                                       ! root to soil flux super-saturating the soils
+                                                       ! This does seem unlikely, but we need accomodate
+                                                       ! small fluxes for various reasons
```
Nice to see this accommodated

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+    real(r8)      , intent(in)     :: pinot
+    real(r8)      , intent(out)    :: dpsi_dth
+
+    dpsi_dth = -1._r8*th_sat*pinot*(rwc_fd - th_res )/((th - th_sat*th_res)**2._r8)
```
Yes, correct

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+    real(r8) , intent(out) :: psi         ! water potential   [MPa]
+
+
+    psi = epsil * (th - th_sat*rwc_fd) / (th_sat*(rwc_fd-th_res)) - pinot
```
Yes, correct.

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+    real(r8)      , intent(in)     :: epsil
+    real(r8)      , intent(out)    :: dpsi_dth       ! derivative of water potential wrt theta  [MPa m3 m-3]
+
+    dpsi_dth = epsil/(th_sat*(rwc_fd - th_res))
```
Yes, correct.

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+    if(psi<this%psi_sat)then
+       dftcdpsi = (-2._r8-3._r8/this%beta) / this%psi_sat * &
+            (psi/this%psi_sat)**(-3._r8-3._r8/this%beta)
+    else
```
Yes, I had same question.

bchristo replied:
Wait, this is not needed -- CCH has no analogue of residual.
Only van Genuchten

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+    ! Differentiate:
+    ! psi = this%psi_sat*(th/this%th_sat)**(-this%beta)
+
+    dpsidth = -this%beta*this%psi_sat/this%th_sat * (th/this%th_sat)**(-this%beta-1._r8)
```
As far as I can see, for CCH, no. There is no residual analogue for CCH

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+    satfrac = th/this%th_sat
+    if(satfrac>max_sf_interp) then
+        psi = this%psi_max + this%dpsidth_max*(th-max_sf_interp*this%th_sat)
+    else
```
Not needed for CCH (see comments below).

rgknox replied:
agreed resolving, no residual available...

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+    if(psi>this%psi_max) then
+        ! Linear range for extreme values
+        th = max_sf_interp*this%th_sat + (psi-this%psi_max)/this%dpsidth_max
+    else
```
See comments below

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+
+    if(th>th_interp) then
+       satfrac = max_sf_interp
+    else
```
Here, for VG, yes, we need this.

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+       psi_interp = -(1._r8/this%alpha)*(max_sf_interp**(1._r8/(m-1._r8)) - 1._r8 )**m
+       psi = psi_interp + dpsidth_interp*(th-th_interp)
+
+    else
```
Agreed

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+       ! convert to volumetric water content
+       th = satfrac*(this%th_sat-this%th_res) + this%th_res
+
+    else
```
agreed.

bchristo commented on biogeophys/FatesHydroWTFMod.F90:
```diff
@@ -0,0 +1,1136 @@
+                                               
+
+  real(r8), parameter :: min_sf_interp = 0.02 ! Linear interpolation above this saturated frac
+  real(r8), parameter :: max_sf_interp = 0.95 ! Linear interpolation below this saturated frac
```
THis comment should read "Linear interpolation above this saturated frac"

rgknox commented:
building a PR on my repo to address review requests:  https://github.com/rgknox/fates/pull/9

rgknox commented:
@xuchongang, @bchristo , @JunyanDing et al.

I think there are some good points brought up in this thread, that we need to have some type of protections in place to help the system work through super dry conditions.  I had originally thought that making the PV curve linear as we approach residual would be helpful, but my take right now is that we need to think about this more.  Making the PV curve linear just short of the residual, will for all concave down PV curves (which is all of them), actually seems more likely that we would overshoot the residual because the pressure remains higher as it approaches residual, and higher pressures would have higher fraction of total conductivity and thus allow more rapid transport in drying conditions.  
I therefore would like us to punt on this, so that we can, as a team, take some time and test different formulations of how to protect against running into the residual.  For instance, perhaps it would be most useful allow the pressure to stay non-linear, but enforce that the FTC reaches zero before we hit residual.

rgknox commented:
[test-hydro-cz2-inv-v6_plots.pdf](https://github.com/NGEET/fates/files/4506321/test-hydro-cz2-inv-v6_plots.pdf)

Here are results from the California CZ2 site, showing some of the new diagnostics.  Note that the SCPF (size-class x pft) diagnostics are just sums over all bins (site sum).

The root water uptake rates for different soil horizons show some redistribution which is cool.  
 FATES_ROOTUPTAKE0_SCPF  is root water uptake from 0-10 cm in the soil, FATES_ROOTUPTAKE10_SCPF is from 10-50 cm, FATES_ROTUPTAKE50_SCPF is from 50-100 and FATES_ROOTUPTAKE100 is from 100+.  

There are also a few variables with the prefix ROOTWGT,  these are site level, and take a mean over the soil column using the fine root surface area as a weighting factor.  For instance one has the volumetric water content, and one has the matric potential, another just shows the saturation value, which is constant.

ckoven commented:
@rgknox is the time axis for the hourly plots in GMT?

rgknox commented:
yes

rgknox commented:
Tests on cheyenne look good:

/glade/scratch/rgknox/clmed-tests/fates-sci.1.35.0_api.10.0.0-clm5.0.30-Ce592cb82-F0651aaec.fates.cheyenne.intel
/glade/scratch/rgknox/clmed-tests/fates-sci.1.35.0_api.10.0.0-clm5.0.30-Ce592cb82-F0651aaec.fates.cheyenne.gnu

Comparisons with baselines don't technically pass, but I checked the diffs on the tests and they were only complaints about different dimension sizes in the output file, which is consistent with changes in the parameter file.
The Size-age mortality tests now pass (because of updated parameter file).  Note these parameter file changes should be backwards compatible, so no changes necessary for users (although they may want to head new defaults for Hydro and Size-age mortality.

Hydro tests now pass as well.

rgknox commented:
retested with changes to fix NEP calculation, expected pass:
/glade/scratch/rgknox/clmed-tests/fates-sci.1.35.0_api.10.0.0-clm5.0.30-C37c1d1b0-F320a8fc0.fates.cheyenne.gnu
/glade/scratch/rgknox/clmed-tests/fates-sci.1.35.0_api.10.0.0-clm5.0.30-C37c1d1b0-F320a8fc0.fates.cheyenne.intel

glemieux commented:
> Comparisons with baselines don't technically pass, but I checked the diffs on the tests and they were only complaints about different dimension sizes in the output file, which is consistent with changes in the parameter file.

Is this `NPLANT_CAPF` and `NPLANT_CACLS`?

glemieux commented on biogeophys/FatesPlantHydraulicsMod.F90:
```diff
@@ -1288,407 +1239,432 @@ end subroutine DeallocateHydrCohort
-          sites(s)%si_hydr%nlevsoi_hyd = bc_in(s)%nlevsoil
-          call sites(s)%si_hydr%InitHydrSite()
+       csite_hydr%nlevrhiz = csite_hydr%i_rhiz_b-csite_hydr%i_rhiz_t+1
+       call sites(s)%si_hydr%InitHydrSite(numpft,nlevsclass)
```
`numpft` is being used for this call but looks like it's not defined as an input to the call anymore.  I noticed this based on the call from the CTSM PR.  Should this be a locally defined variable?

rgknox replied:
It is a global: https://github.com/NGEET/fates/blob/320a8fc0875e228df4133544f122c83844d8720e/biogeophys/FatesPlantHydraulicsMod.F90#L62

glemieux replied:
doh :facepalm: .  I shoulda seen that.

glemieux commented:
Cross tagging #633 for future reference.

bchristo commented:
@rgknox I looked at the diagnostic plots for Ca CZ2 site and they look
pretty good to me. I paid close attention to QVEGT and Sapflow and it does
look like at the diurnal amplitude of QVEGT is greater than Sapflow, which
is what I would expect since sapflow responds only after internal plant
storage is drawn upon.  A second thing which would be helpful would be to
see the diurnal plots split into wet vs. dry season, or alternatively, have
history variables for predawn and midday water potential and plot those
seasonally - maybe we add those two variables to the current history output
wish list?

On Mon, Apr 20, 2020 at 4:50 PM Ryan Knox <notifications@github.com> wrote:

> yes
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NGEET/fates/pull/611#issuecomment-616829460>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADSXQZSQUE4DY3C2RTV4TNTRNS7Q3ANCNFSM4KXLFY6A>
> .
>
